### PR TITLE
Implement a `--show-source` setting

### PR DESCRIPTION
**Pull request by harupy**
See #525. 

```
cargo run -- --no-cache --select C resources/test/fixtures/ --show-source
```

image

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -13,6 +13,7 @@ edition = "2021"
 name = "ruff"
 
 [dependencies]
+annotate-snippets = { version = "0.9.1", features = ["color"] }
```
Comment (harupy) [inline Cargo.toml:16]: rustc uses this crate: https://github.com/rust-lang/rust/issues/59346#issuecomment-1279325650

Comment (harupy) [reply Cargo.toml:16]: Another option is https://github.com/zkat/miette. SWC uses this uses this crate https://github.com/swc-project/swc/blob/27225a0eeb78f3c2975ec14ef51127b2b9532a84/crates/swc_error_reporters/Cargo.toml#L16

**Comment by charliermarsh** [comment]
\cc @squiddy

**Comment by charliermarsh** [comment]
Wow nice. Is it possible to include the fix annotations / suggestions here too, like Clippy does?

**Comment by harupy** [comment]
That's what I'm investigating now.

File: Cargo.toml (hunk 1 above)
Comment (charliermarsh) [reply Cargo.toml:16]: It looks like that uses https://crates.io/crates/ariadne under the hood.

**Comment by charliermarsh** [comment]
I do like the look and feel! And I love not implementing this ourselves.

**Comment by harupy** [comment]
@charliermarsh It looks like clippy relies on `rustc_errors` (rustc's private crate) for creating an error report.

File: src/message.rs [hunk 2]
```diff
@@ -62,3 +64,57 @@ impl fmt::Display for Message {
...
+        let code = self.kind.code().as_ref();
+        let rel_path = relativize_path(Path::new(&self.filename));
+        let source = error_lines.join("\n");
+        let snippet = Snippet {
+            title: Some(Annotation {
+                label: Some(&body),
+                id: Some(code),
+                annotation_type: AnnotationType::Error,
+            }),
+            footer: vec![],
```
Comment (harupy) [inline src/message.rs:101]: This is where we can show notes or help messages:

```diff
-            footer: vec![],
+            footer: vec![Annotation {
+                label: Some("Add --fix to automatically this error"),
+                id: None,
+                annotation_type: AnnotationType::Help,
+            }],
```

image

**Comment by charliermarsh** [comment]
Could we vendor that crate? Or is it too closely coupled to Clippy / rustc to be extracted?

**Comment by harupy** [comment]
This is `rustc_errors`' `Cargo.toml`:
https://github.com/rust-lang/rust/blob/fb6667a233fa677f6fbfa6067913ef3d32480317/compiler/rustc_errors/Cargo.toml

As you can see, it depends on other private crates:

```toml
[package]
name = "rustc_errors"
version = "0.0.0"
edition = "2021"

[lib]

[dependencies]
tracing = "0.1"
rustc_ast = { path = "../rustc_ast" }
rustc_ast_pretty = { path = "../rustc_ast_pretty" }
rustc_error_messages = { path = "../rustc_error_messages" }
rustc_serialize = { path = "../rustc_serialize" }
rustc_span = { path = "../rustc_span" }
rustc_macros = { path = "../rustc_macros" }
rustc_data_structures = { path = "../rustc_data_structures" }
rustc_target = { path = "../rustc_target" }
rustc_hir = { path = "../rustc_hir" }
rustc_lint_defs = { path = "../rustc_lint_defs" }
unicode-width = "0.1.4"
atty = "0.2"
termcolor = "1.0"
annotate-snippets = "0.9"
termize = "0.1.1"
serde = { version = "1.0.125", features = [ "derive" ] }
serde_json = "1.0.59"

[target.'cfg(windows)'.dependencies]
winapi = { version = "0.3", features = [ "handleapi", "synchapi", "winbase" ] }
```

File: src/message.rs [hunk 3]
```diff
@@ -62,3 +64,57 @@ impl fmt::Display for Message {
...
+                .map(|(i, line)| {
+                    if i == error_lines.len() - 1 {
+                        self.end_location.column()
+                    } else {
+                        line.len() + 1
+                    }
+                })
+                .sum::<usize>()
+                - 1,
+        );
```
Comment (charliermarsh) [inline src/message.rs:94]: Can we find a way to do all of the above through the cached `SourceCodeLocator` that we already create for each file? (Maybe it involves adding the body text to `message` or something?)

File: src/message.rs [hunk 4]
```diff
@@ -62,3 +64,57 @@ impl fmt::Display for Message {
...
+                    annotation_type: AnnotationType::Error,
+                    range,
+                }],
+            }],
+            opt: FormatOptions {
+                color: true,
+                ..Default::default()
+            },
+        };
+        DisplayList::from(snippet).to_string()
```
Comment (charliermarsh) [inline src/message.rs:118]: I think that if we're going to use this crate for `--show-source`, we should probably use the same crate and style even when `--show-source` is false (and just omit the source code). What do you think?

File: src/message.rs (hunk 3 above)
Comment (harupy) [reply src/message.rs:94]: > Maybe it involves adding the body text to message or something?

I'll try this.

File: src/message.rs (hunk 4 above)
Comment (harupy) [reply src/message.rs:118]: Sounds good. I think we can set `slices` to an empty vector when `--show-source` is false.

**Comment by harupy** [comment]
Updated the screenshot.

File: src/cli.rs [hunk 5]
```diff
@@ -94,6 +94,9 @@ pub struct Cli {
     /// The name of the file when passing it through stdin.
     #[arg(long)]
     pub stdin_filename: Option<String>,
+    /// Show violations with source code.
+    #[arg(long)]
+    pub show_source: bool,
```
Comment (charliermarsh) [inline src/cli.rs:99]: I think we should add this to `pyproject.toml` too, wdyt?

File: src/message.rs [hunk 6]
```diff
@@ -16,16 +20,40 @@ pub struct Message {
...
+        let source = locator.slice_source_code_range(&Range {
+            location: Location::new(check.location.row(), 0),
+            end_location: Location::new(check.end_location.row() + 1, 0),
+        });
+        let num_chars_in_range = locator
+            .slice_source_code_range(&Range {
+                location: check.location,
+                end_location: check.end_location,
+            })
+            .len();
```
Comment (charliermarsh) [inline src/message.rs:44]: I think this needs to be `.chars().len()`, or we need to use the `Rope` in `locator` to get the start and end _character_ positions.

File: src/message.rs [hunk 7]
```diff
@@ -58,7 +85,42 @@ impl fmt::Display for Message {
             self.location.column(),
             ":".cyan(),
             self.kind.code().as_ref().red().bold(),
-            self.kind.body()
-        )
+            self.kind.body(),
+        );
+        let slices = if self.show_source && self.source.is_some() && self.range.is_some() {
+            vec![Slice {
+                source: self.source.as_ref().unwrap(),
```
Comment (charliermarsh) [inline src/message.rs:92]: Nit: maybe `if let` to avoid some of the unwraps?

File: src/cli.rs (hunk 5 above)
Comment (harupy) [reply src/cli.rs:99]: Looks like I forgot to update `user.rs`. Updated it.

File: src/message.rs (hunk 7 above)
Comment (harupy) [reply src/message.rs:92]: Sounds good. Will fix.

**Comment by charliermarsh** [comment]
Sweet! Will try to merge this tomorrow (or maybe the next day -- I have to move apartments tomorrow so might be busy). I just want to do some performance testing before merging.

**Comment by charliermarsh** [comment]
(Sorry, I'm a little behind, I haven't had a chance to benchmark yet.)

**Comment by harupy** [comment]
@charliermarsh Not a problem at all :)

**Comment by harupy** [comment]
@charliermarsh Thanks for the clean-up and merge! This does help when verifying new lint rules.

**Pull request merged**